File: Programmers/[72415] 카드 짝 맞추기.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    static final int INF = 987654321;
    static final int[] dy = {-1, 0, 1, 0};
    static final int[] dx = {0, 1, 0, -1};
    static List<Integer> answer;

    static int calculatePath(int fromY, int fromX, int toY, int toX, List<Card> cards) {
        // 8가지 방향
        int[][] visited = new int[4][4];
        for (int[] v : visited) {
            Arrays.fill(v, INF);
        }
        Queue<Position> queue = new LinkedList<>();

        queue.offer(new Position(fromY, fromX));
        visited[fromY][fromX] = 0;

        while (!queue.isEmpty()) {
            Position cur = queue.poll();
            if (cur.y == toY && cur.x == toX) {
                return visited[toY][toX];
            }

            // just one
            for (int i = 0; i < 4; i++) {
                int ny = cur.y + dy[i];
                int nx = cur.x + dx[i];
                if (0 <= ny && ny < 4 && 0 <= nx && nx < 4) {
                    if (visited[ny][nx] == INF) {
                        visited[ny][nx] = visited[cur.y][cur.x] + 1;
                        queue.offer(new Position(ny, nx));
                    }
                }
            }

            // + ctrl
            for (int i = 0; i < 4; i++) {
                int length = 0;
                boolean isPossible = true;
                while (isPossible) {
                    int ny = cur.y + (length + 1) * dy[i];
                    int nx = cur.x + (length + 1) * dx[i];

                    if (ny < 0 || ny >= 4 || nx < 0 || nx >= 4) {
                        break;
                    }
                    for (Card card : cards) {
                        if (card.y == ny && card.x == nx) {
                            isPossible = false;
                            break;
                        }
                    }

                    length++;
                }

                int ny = cur.y + length * dy[i];
                int nx = cur.x + length * dx[i];
                if (visited[ny][nx] == INF) {
                    visited[ny][nx] = visited[cur.y][cur.x] + 1;
                    queue.offer(new Position(ny, nx));
                }
            }
        }

        return visited[toY][toX];
    }

    static void dfs(int y, int x, int cnt, List<Card> cards) {
        if (cards.isEmpty()) {
            answer.add(cnt);
            return;
        }

        for (Card card : cards) {
            Card oppositeCard = card.pair;
            int path1 = calculatePath(y, x, card.y, card.x, cards) + 1;
            int path2 = calculatePath(card.y, card.x, oppositeCard.y, oppositeCard.x, cards) + 1;

            List<Card> removedCards = cards.stream()
                    .filter(c -> !c.equals(card) && !c.equals(oppositeCard))
                    .collect(Collectors.toCollection(ArrayList::new));

            dfs(oppositeCard.y, oppositeCard.x, cnt + path1 + path2, removedCards);
        }
    }

    public int solution(int[][] board, int r, int c) {
        answer = new ArrayList<>();

        List<Card> cards = new ArrayList<>();
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                int num = board[i][j];
                if (num != 0) {
                    Card newCard = new Card(num, i, j);
                    for (Card card : cards) {
                        if (card.num == num) {
                            card.setPair(newCard);
                            newCard.setPair(card);
                        }
                    }
                    cards.add(newCard);
                }
            }
        }

        dfs(r, c, 0, cards);

        return Collections.min(answer);
    }

    static class Position {
        int y;
        int x;

        public Position(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }

    static class Card {
        int num;
        int y;
        int x;
        Card pair;

        public Card(int num, int y, int x) {
            this.num = num;
            this.y = y;
            this.x = x;
        }

        void setPair(Card pair) {
            this.pair = pair;
        }
    }
}

File: Programmers/[72415] 카드 짝 맞추기.py
from collections import deque

answer = []
dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]

# BFS
def calculatePath(board, startY, startX, goalY, goalX):
    isVisited = [[-1] * 4 for _ in range(4)]
    isVisited[startY][startX] = 0
    queue = deque([[startY, startX]])
    while queue:
        y, x = queue.popleft()
        if y == goalY and x == goalX:
            return isVisited[y][x]

        for i in range(4):
            # 한 칸 이동
            ny = y + dy[i]
            nx = x + dx[i]
            if 0 <= ny < 4 and 0 <= nx < 4:
                if isVisited[ny][nx] == -1:
                    queue.append([ny, nx])
                    isVisited[ny][nx] = isVisited[y][x] + 1
            # Ctrl 이동
            ny = y
            nx = x
            while True:
                ny += dy[i]
                nx += dx[i]
                if not (0 <= ny < 4 and 0 <= nx < 4):
                    ny -= dy[i]
                    nx -= dx[i]
                    break
                if board[ny][nx] != 0:
                    break
            if isVisited[ny][nx] == -1:
                queue.append([ny, nx])
                isVisited[ny][nx] = isVisited[y][x] + 1
    return -1


def dfs(board, y, x, count, cards):
    if len(list(filter(lambda x: not x.isDeleted, cards))) == 0:
        answer.append(count)
        return

    for card in cards:
        if card.isDeleted:
            continue

        pairCard = card.pair
        path1 = calculatePath(board, y, x, card.y, card.x)
        path2 = calculatePath(board, card.y, card.x, pairCard.y, pairCard.x)
        if path1 == -1 or path2 == -1:
            continue

        cardNumber = board[card.y][card.x]
        board[card.y][card.x] = 0
        board[pairCard.y][pairCard.x] = 0
        card.delete()
        pairCard.delete()

        newCount = count + path1 + path2 + 2  # +2: enter
        dfs(board, pairCard.y, pairCard.x, newCount, cards)

        board[card.y][card.x] = cardNumber
        board[pairCard.y][pairCard.x] = cardNumber
        card.restore()
        pairCard.restore()


def solution(board, r, c):
    cards = []
    for i in range(4):
        for j in range(4):
            if board[i][j] != 0:
                newCard = Card(i, j, board[i][j])
                for card in cards:
                    if card.number == newCard.number:
                        newCard.setPair(card)
                        card.setPair(newCard)
                cards.append(newCard)

    dfs(board, r, c, 0, cards)

    return min(answer)


class Card:
    def __init__(self, y, x, number):
        self.y = y
        self.x = x
        self.number = number
        self.isDeleted = False

    def setPair(self, card):
        self.pair = card

    def delete(self):
        self.isDeleted = True

    def restore(self):
        self.isDeleted = False
